src/Common/src/Microsoft/Win32/SafeHandles/GssSafeHandles.cs

```diff
@@ -0,0 +1,158 @@
+        private SafeGssBufferHandle() : base(IntPtr.Zero, true)
+        {
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        internal void Copy(byte[] destination, int offset)
+        {
+            Debug.Assert(destination != null, "destination buffer cannot be null");
+            Debug.Assert(offset <= destination.Length , "offset in the destination buffer must be valid");
+            if (destination.Length > 0)
+            {
+                Interop.NetSecurityNative.CopyBuffer(this, destination, capacity: destination.Length, offset: offset);
+            }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseBuffer(out minorStatus, handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_name_t_desc*
+    /// </summary>
+    internal sealed class SafeGssNameHandle : SafeHandle
+    {
+        public static SafeGssNameHandle Create(string name, bool isUser)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name), "Invalid name passed to SafeGssNameHandle create");
+            SafeGssNameHandle retHandle;
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = isUser?
+                Interop.NetSecurityNative.ImportUserName(out minorStatus, name, name.Length, out retHandle) :
+                Interop.NetSecurityNative.ImportPrincipalName(out minorStatus, name, name.Length, out retHandle);
+            if (status != Interop.NetSecurityNative.Status.GSS_S_COMPLETE)
+            {
+                throw new Interop.NetSecurityNative.GssApiException(status, minorStatus);
+            }
+
+            return retHandle;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurityNative.Status minorStatus;
+            Interop.NetSecurityNative.Status status = Interop.NetSecurityNative.ReleaseName(out minorStatus, ref handle);
+            SetHandle(IntPtr.Zero);
+            return status == Interop.NetSecurityNative.Status.GSS_S_COMPLETE;
+        }
+
+        private SafeGssNameHandle()
+            : base(IntPtr.Zero, true)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a gss_cred_id_t_desc_struct*
+    /// </summary>
+    internal class SafeGssCredHandle : SafeHandle
+    {
+        public static SafeGssCredHandle Create(string username, string password, string domain)
+        {
+            SafeGssCredHandle retHandle = null;
+
+            // Empty username is OK if Kerberos ticket was already obtained
```

Comment:
In that case it's going to return a null SafeGssCredHandle.  That's not exactly what I was expecting from this comment.  If returning null here is correct, consider updating the comment to indicate how the null is expected to be handled.